### ISSUE TITLE
zzykxx - Vaults are not boostrapped atomically

### ISSUE THREAD

**Issue by sherlock-admin3** [sherlock-audit/2024-03-amphor-judging#50]
zzykxx

medium

# Vaults are not boostrapped atomically

## Vulnerability Detail
The protocol team mentioned that vaults will be bootstrapped with initial liquidity to prevent inflation attacks. To effectively prevent an inflation attack the funds should be deposited in the vault atomically during [initialization](https://github.com/sherlock-audit/2024-03-amphor/blob/main/asynchronous-vault/src/AsyncSynthVault.sol#L181), but this doesn't happen, opening up the possibility of an inflation attack on the bootstrapping transaction itself.

## Impact
The protocol team could fall victim to an inflation attack on their first deposit.

## Tool used

Manual Review

## Recommendation
Make the first deposit in the vault during the [initialization](https://github.com/sherlock-audit/2024-03-amphor/blob/main/asynchronous-vault/src/AsyncSynthVault.sol#L181) call, this way the first deposit can't be front-run.

**Comment by sherlock-admin4** [comment]
The protocol team fixed this issue in the following PRs/commits:
https://github.com/AmphorProtocol/asynchronous-vault/pull/103

**Comment by realfugazzi** [comment]
I guess someone should escalate this one, when the protocol team stated they will bootstrap vaults it should be assumed they will be handling this through other on-chain (factories) or off-chain (deployment scripts with checks before launch) mechanisms not in scope for the current contest.

The current issue speculates this should be handled in the initialization of the vault, which should not be assumed, and ultimately is not fair for other participants.

cc @WangSecurity

**Comment by 0xLogos** [comment]
Escalate

Agree with comment above

**Comment by sherlock-admin2** [comment]
> Escalate
> 
> Agree with comment above

You've created a valid escalation!

To remove the escalation from consideration: Delete your comment.

You may delete or edit your escalation comment anytime before the 48-hour escalation window closes. After that, the escalation becomes final.

**Comment by Arabadzhiew** [comment]
Escalate

I disagree with the above escalation. How are we supposed to "assume" that the protocol team will handle the bootstrap process safely when we do not have any of the supposed factories, deployment scripts, etc. in the scope of the contest? If we go down that route, we might as well "assume" that the transactions sent to the protocol will go through a private mempool and eliminate the possibility of any front/back-running vulnerabilities altogether.

I would go even further and argue that this issue should be judged as a high severity one, instead of medium. The inflation vulnerability here opens the possibility for the first depositor (the protocol team in our case) to receive 0 shares in return for their deposited assets, which basically means that they can get all of their deposited funds stollen. Issue #129 goes into more detail as to how that is possible and provides a step-by-step example of a basic exploit scenario.

**Comment by sherlock-admin2** [comment]
> Escalate
> 
> I disagree with the above escalation. How are we supposed to "assume" that the protocol team will handle the bootstrap process safely when we do not have any of the supposed factories, deployment scripts, etc. in the scope of the contest? If we go down that route, we might as well "assume" that the transactions sent to the protocol will go through a private mempool and eliminate the possibility of any front/back-running vulnerabilities altogether.
> 
> I would go even further and argue that this issue should be judged as a high severity one, instead of medium. The inflation vulnerability here opens the possibility for the first depositor (the protocol team in our case) to receive 0 shares in return for their deposited assets, which basically means that they can get all of their deposited funds stollen. Issue #129 goes into more detail as to how that is possible and provides a step-by-step example of a basic exploit scenario. 

You've created a valid escalation!

To remove the escalation from consideration: Delete your comment.

You may delete or edit your escalation comment anytime before the 48-hour escalation window closes. After that, the escalation becomes final.

**Comment by WangSecurity** [comment]
What fugazzi said above is actually what made me believe this issue is invalid, initially. But during internal discussions with the sponsor they said they wanted to update the code for the automatic bootstrap, but they didn't make it on time before the start of the contest. Therefore, when LSW brought up this issue, sponsor agreed this one should be valid.

**Comment by realfugazzi** [comment]
> What fugazzi said above is actually what made me believe this issue is invalid, initially. But during internal discussions with the sponsor they said they wanted to update the code for the automatic bootstrap, but they didn't make it on time before the start of the contest. Therefore, when LSW brought up this issue, sponsor agreed this one should be valid.

That's fine, but judging should be done based on the information we all had. Even the docs state the issue is known and should not result in a valid finding.

https://github.com/sherlock-audit/2024-03-amphor?tab=readme-ov-file#q-please-list-any-known-issuesacceptable-risks-that-should-not-result-in-a-valid-finding

**Comment by 0xCryptoSan** [comment]
I think this issue is invalid. I didn't report this issue because, in the contest description, it was clearly written that the team would bootstrap the vault. In my eyes, it is not fair to say now that we cannot assume the team will bootstrap the vault and make an inflation attack a valid finding.

**Comment by WangSecurity** [comment]
The contest's README says: "We will boostrap the vault to avoid inflation attack so please take this into consideration". It doesn't say that this shouldn't result in a valid finding (I know the question says "Please list any known issues/acceptable risks that should not result in a valid finding", but the protocol said, that this should be "taken into consideration").
And I think these two watsons took it into consideration. I believe this should remain valid since in the code implementation there was no signs of bootstrapping. Therefore, as both of these reports (50 and 129) say, the team would become the victim in that case.

**Comment by aslanbekaibimov** [comment]
The core of the attack is to make 1 share worth more than the bootstrap deposit, but because there's 1 virtual share the attacker has to donate at least 2x the bootstrap deposit (1x for his share and 1x for the virtual share).

Then the bootstrap deposit will be absorbed by attacker's share and the virtual share, so each will become worth 1.5x the bootstrap deposit. So the attacker donates 2x and takes out only 1.5x, in other words he loses 500 USDC in this case.

The only thing the attacker can do is donate more, so the loss for him becomes smaller (because there will be more attacker's shares absorbing the victim's deposit):

E.g. if the budget is 10k USDC, the attacker would mint 9 shares and donate 10k, making every share worth `1000 USDC + dust`. After the bootstrap txn, the 10 shares (9 attacker's and 1 virtual) will be worth 1100 USDC each. Attacker's loss = 10000 USDC - 9900 USDC = 100 USDC. The loss for the attacker shrinks with the budget, but it would never be a gain.

Moreover, funds absorbed by the virtual share can be recovered via proxy upgrade.

TLDR

1. Attacker has to have a budget (> 2000 USDC), is willing to lose funds and to spend gas
2. The only victim is the protocol
3. Funds are recoverable

**Comment by realfugazzi** [comment]
> The contest's README says: "We will boostrap the vault to avoid inflation attack so please take this into consideration". It doesn't say that this shouldn't result in a valid finding (I know the question says "Please list any known issues/acceptable risks that should not result in a valid finding", but the protocol said, that this should be "taken into consideration"). And I think these two watsons took it into consideration. I believe this should remain valid since in the code implementation there was no signs of bootstrapping. Therefore, as both of these reports (50 and 129) say, the team would become the victim in that case.

The "take this into consideration" is in direct reference to the title of the question there. The protocol is filling a section that informs participants of things that won't be taken as valid issues.

As mentioned, the bootstrap doesn't necessarily need to happen in the assets in scope or on-chain at all.

**Comment by WangSecurity** [comment]
Agree with what Aslanbek said above and also checked all the calculations myself. At this point it looks like the attacker won't be able to extract all the funds due to increments in the formula (`asset.mulDiv(totalSupply() + 1, totalAssets() + 1, rounding)`. If the attacker deposits 2000 USDC when the bootrstrapping is 1000 USDC, they will withdraw only 1500 USDC, so it's a loss. Yes they can deposit more and be in profit (as I understand), the bootrstrapper will receive no shares, but there's a functionality to recover all the funds from the protocol.

As for what fugazzi said, this argument can be used by both sides. As I've said this doesn't mean all such vulns should be taken out of concern cause "take it into consideration" is a bit vague. But it also a valid argument that we should assume it can happen in other out-of-scope contracts (hope we both agree that's valid arguments for both sides).

Therefore, since "take it consideration" is quite vague and can mean lots of things, and I believe it should stay valid due to the following reasons:

1. The sponsor agreed it's a valid finding.
2. It was a risky submission, cause we can easily consider the bootstrapping to be done in another way, not in the in-scope contract, but the watsons still decided to submit and risk their valid/invalid ratio.

I know it's not the perfect arguments, but I can see how both of sides can be right (both the ones who escalated and me), therefore, I will take any decision from the head of judging.

Hope you also see and understand my points. And big thank you to Aslanbek and Fugazzi for bringing up these points, they're very insightful!

**Comment by realfugazzi** [comment]
> Agree with what Aslanbek said above and also checked all the calculations myself. At this point it looks like the attacker won't be able to extract all the funds due to increments in the formula (`asset.mulDiv(totalSupply() + 1, totalAssets() + 1, rounding)`. If the attacker deposits 2000 USDC when the bootrstrapping is 1000 USDC, they will withdraw only 1500 USDC, so it's a loss. Yes they can deposit more and be in profit (as I understand), the bootrstrapper will receive no shares, but there's a functionality to recover all the funds from the protocol.
> 
> As for what fugazzi said, this argument can be used by both sides. As I've said this doesn't mean all such vulns should be taken out of concern cause "take it into consideration" is a bit vague. But it also a valid argument that we should assume it can happen in other out-of-scope contracts (hope we both agree that's valid arguments for both sides).
> 
> Therefore, since "take it consideration" is quite vague and can mean lots of things, and I believe it should stay valid due to the following reasons:
> 
>     1. The sponsor agreed it's a valid finding.
> 
>     2. It was a risky submission, cause we can easily consider the bootstrapping to be done in another way, not in the in-scope contract, but the watsons still decided to submit and risk their valid/invalid ratio.
> 
> 
> I know it's not the perfect arguments, but I can see how both of sides can be right (both the ones who escalated and me), therefore, I will take any decision from the head of judging.
> 
> Hope you also see and understand my points. And big thank you to Aslanbek and Fugazzi for bringing up these points, they're very insightful!

We are not saying the issue is invalid, but questioning if it's fair to be accepted when your own rules say it should not be valid. 
In any case, have you also checked the validity of the issue? There's no PoC and the vault is based on OZ implementation which mitigates the inflation attack.

**Comment by WangSecurity** [comment]
Regarding your first sentence, I wanted to highlight that this specific case is not that clear (at least for me), cause "take into consideration" sounds ambiguous.
As for your question, I've made several calculations for all other issues, except this one :), cause when I looked at it and the code, and it seemed valid for me, even tho there are `+1` for assets and shares.

I've just did the calculations and yeah, the problem is that nobody gains or loses anything. If we bootstrap the vault with 1000 USDC and attacker tries inflation attack with 1 wei of shares and 1000 USDC, then when they try to withdraw they will only get 1000 USDC.

And the larger budget they use, the more they lose. If they use 2000 USDC, they then withdraw only 1500 USDC, and the vault has remaining 150 USDC. If they use 10k, then they get out 5000 USDC in the end.

Yep, the inflation attack also seems impossible right now, maybe I'm missing again something here, but at this point I agree with the escalation, since the inflation attack seems pointless (either no profit or even loss).

**Comment by goheesheng** [comment]
It's not possible to have an inflation attack. For OP, prove with PoC then.

**Comment by zzykxx** [comment]
> The core of the attack is to make 1 share worth more than the bootstrap deposit, but because there's 1 virtual share the attacker has to donate at least 2x the bootstrap deposit (1x for his share and 1x for the virtual share).
> 
> Then the bootstrap deposit will be absorbed by attacker's share and the virtual share, so each will become worth 1.5x the bootstrap deposit. So the attacker donates 2x and takes out only 1.5x, in other words he loses 500 USDC in this case.
> 
> The only thing the attacker can do is donate more, so the loss for him becomes smaller (because there will be more attacker's shares absorbing the victim's deposit):
> 
> E.g. if the budget is 10k USDC, the attacker would mint 9 shares and donate 10k, making every share worth `1000 USDC + dust`. After the bootstrap txn, the 10 shares (9 attacker's and 1 virtual) will be worth 1100 USDC each. Attacker's loss = 10000 USDC - 9900 USDC = 100 USDC. The loss for the attacker shrinks with the budget, but it would never be a gain.
> 
> Moreover, funds absorbed by the virtual share can be recovered via proxy upgrade.
> 
> TLDR
> 
> 1. Attacker has to have a budget (> 2000 USDC), is willing to lose funds and to spend gas
> 2. The only victim is the protocol
> 3. Funds are recoverable

This is correct. There is an edge case where an attacker inflates the value of shares, the bootstrap team deposits, and then other users deposit which could potentially make the attack profitable. This being said, I think this should be low severity at best.

**Comment by Evert0x** [comment]
First of all, I would interpret the README that the inflation attack was labeled as a known issue by the protocol team, which would make this issue invalid.

> **Please list any known issues/acceptable risks that should not result in a valid finding.**
> We will boostrap the vault to **avoid inflation attack** so please take this into consideration.
> For weth vault it would be 1 weth.
> For usdc vault it would be 1000 usdc.

But even if interpreted otherwise, the issue submitter itself agrees the report should be invalidated @zzykxx.

Planning to accept escalation and invalidate issue.

**Comment by Evert0x** [comment]
Result:
Invalid 
Has Duplicates

**Comment by sherlock-admin3** [comment]
Escalations have been resolved successfully!

Escalation status:
- [0xLogos](https://github.com/sherlock-audit/2024-03-amphor-judging/issues/50/#issuecomment-2025348039): accepted
- [Arabadzhiew](https://github.com/sherlock-audit/2024-03-amphor-judging/issues/50/#issuecomment-2025443223): rejected

**Comment by sherlock-admin3** [comment]
The Lead Senior Watson signed off on the fix.
